### DISCUSSION TITLE
coc extension contains C code cannot be build when `CocInstall`

### DISCUSSION THREAD

**Discussion by Freed-Wu** [neoclide/coc.nvim#4960] [Q&A] [answered]
For a nodejs project containing `*.gyp` and `*.c`, `npm pack` and `npm install` will build `*.c` by [`node-gyp`](https://github.com/nodejs/node-gyp). However, `CocInstall` will not build `*.c`. Why?

[Referred example](https://github.com/tonyfettes/coc-rime)

**Comment by fannheyward** [comment, marked as answer]
> If there is a binding.gyp file in the root of your package and you haven't defined your own install or preinstall scripts, npm will default the install command to compile using node-gyp via node-gyp rebuild 
> https://docs.npmjs.com/cli/v10/using-npm/scripts#npm-install

`npm install` runs `node-gyp rebuild` automatically in project that contains native add-ons `binding.gyp`

CocInstall steps:

1. download and extract extension tar file from https://registry.npmjs.org to extensions folder `~/.config/coc/extensions/node_modules`, with an uniq temp folder
2. the extension contains bundled `lib/index.js` file, maybe with some dependencies
3. `npm install --ignore-scripts --no-lockfile --omit=dev --legacy-peer-deps --no-global` to install the dependencies
4. the `--ignore-scripts` ignores all scripts, includes `postinstall`, won't trigger `node-gyp rebuild` to build native add-ons

**Reply by Freed-Wu** [reply]
So we must build `binding.gyp` manually after `CocInstall`, right?

**Reply by fannheyward** [reply]
I'd suggest to build and publish the compiled native binaries. I cloned the coc-rime, run `npm install` but failed

```
Package rime was not found in the pkg-config search path.
Perhaps you should add the directory containing `rime.pc'
to the PKG_CONFIG_PATH environment variable
No package 'rime' found
Package json-c was not found in the pkg-config search path.
Perhaps you should add the directory containing `json-c.pc'
to the PKG_CONFIG_PATH environment variable
No package 'json-c' found
```

**Reply by Freed-Wu** [reply]
> publish the compiled native binaries

Even if publish the compiled native binaries, we still need user to install the library because it is linked dynamically. Maybe we must encourage user to build it by themselves `npm rebuild`. Or download static linked program.

**Reply by Freed-Wu** [reply]
1. Why postinstall is disabled by coc? Is it due to performance?
2. Is it possible to allow postinstall for coc extension?

**Reply by fannheyward** [reply]
Check out coc-mocword, use mocword from PATH, users must install the binary manually.

**Reply by fannheyward** [reply]
1. `ignore-scripts` to prevent unexpected behaviors, postinstall can be set any commands
2. There's no way to allow extensions to run postinstall for now
3. My thought: instead of allowing users to build but failed, it's better to tell users to install manually.

**Reply by Freed-Wu** [reply]
> it's better to tell users to install manually.

Sometimes the binary file is not a binary file can be installed seperatedly, For example, a library depends on `node.h` or `nan.h`. So is any project to allow extensions to run postinstall in futurn?

**Reply by fannheyward** [reply]
No plan yet.

**Reply by Freed-Wu** [reply]
OK, before this feature be supported, let user `cd ~/.config/coc/extensions/node_modules/coc-XXX && npm rebuild` by themselves.

Thanks for your answer! :+1:

**Reply by fannheyward** [reply]
Another solution: use plugin manager, like vim-plug, to install extensions with build scripts. 

`Plug 'tonyfettes/coc-rime', {'do': 'yarn install --frozen-lockfile'}`

**Reply by Freed-Wu** [reply]
Interesting. If a coc extension depends on a package which have a `binding.gyp`, this dependency will not be build, too, Right?

**Reply by fannheyward** [reply]
I think it should be built. Install extensions via plugins manager will 1. clone extension files to vim runtime folder, 2. run something like `npm install` by `do` command, same as all you did in local.

**Reply by Freed-Wu** [reply]
> Interesting. If a coc extension depends on a package which have a binding.gyp, this dependency will not be build, too, Right?

Oh, I means `CocInstall` a coc extension depending on a package containing binding.gyp. Will it be built?

**Reply by fannheyward** [reply]
Haven't do a test for this case, I think the dependency probably not be built.

**Reply by Freed-Wu** [reply]
Another method: run `exec('npm rebuild')` in Coc extension: https://github.com/tonyfettes/coc-rime/blob/master/src/config.ts#L56

and tell user how to install dependencies in README.md https://github.com/tonyfettes/coc-rime#dependencies

**Reply by Freed-Wu** [reply]
Perhaps https://github.com/prebuild/prebuildify can be helpful.

**Reply by Freed-Wu** [reply]
Related PR: #5002
